File: src/main/java/io/github/intellij/dlanguage/actions/ProcessDLibs.java
package io.github.intellij.dlanguage.actions;

import com.intellij.execution.ExecutionException;
import com.intellij.execution.configurations.GeneralCommandLine;
import com.intellij.execution.configurations.ParametersList;
import com.intellij.execution.process.OSProcessHandler;
import com.intellij.execution.process.ProcessAdapter;
import com.intellij.execution.process.ProcessEvent;
import com.intellij.notification.Notification;
import com.intellij.notification.NotificationType;
import com.intellij.notification.Notifications;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.application.ModalityState;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.module.ModuleManager;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.progress.Task.Backgroundable;
import com.intellij.openapi.progress.impl.BackgroundableProcessIndicator;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.roots.LibraryOrderEntry;
import com.intellij.openapi.roots.ModifiableModelsProvider;
import com.intellij.openapi.roots.ModifiableRootModel;
import com.intellij.openapi.roots.ModuleRootModificationUtil;
import com.intellij.openapi.roots.OrderRootType;
import com.intellij.openapi.roots.impl.OrderEntryUtil;
import com.intellij.openapi.roots.libraries.Library;
import com.intellij.openapi.roots.libraries.LibraryTable;
import com.intellij.openapi.roots.libraries.LibraryTablesRegistrar;
import com.intellij.openapi.ui.popup.JBPopup;
import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.openapi.util.Key;
import com.intellij.openapi.util.SystemInfo;
import com.intellij.openapi.vfs.LocalFileSystem;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.VirtualFileManager;
import com.intellij.ui.components.JBList;
import io.github.intellij.dlanguage.module.DlangModuleType;
import io.github.intellij.dlanguage.project.DubConfigurationParser;
import io.github.intellij.dlanguage.project.DubPackage;
import io.github.intellij.dlanguage.settings.ToolKey;
import io.github.intellij.dlanguage.utils.DToolsNotificationListener;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import javax.swing.JList;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class ProcessDLibs extends AnAction implements DumbAware {

    public static final String MENU_PATH = "Tools > Process D Libraries";
    private static final String NOTIFICATION_GROUPID = "Process D Libs";
    private static final Logger LOG = Logger.getInstance(ProcessDLibs.class);

    private static boolean enabled(@NotNull final AnActionEvent e) {
        final Project project = getEventProject(e);
        if (project == null) {
            return false;
        }
        final String dubPath = ToolKey.DUB_KEY.getPath();
        return dubPath != null && !dubPath.isEmpty()
            && DlangModuleType.findModules(project).size() > 0;
    }

    private static void showModuleChoicePopup(@NotNull final AnActionEvent e, final Project project,
        final Collection<Module> modules) {
        final JList list = new JBList(JBList.createDefaultListModel(modules.toArray()));
        final JBPopup popup = JBPopupFactory.getInstance()
            .createListPopupBuilder(list)
            .setTitle("Process D libraries for module")
            .setItemChoosenCallback(makeModuleChoiceCallback(e, list))
            .createPopup();
        popup.showCenteredInCurrentWindow(project);
    }

    private static Runnable makeModuleChoiceCallback(final @NotNull AnActionEvent event,
        final @NotNull JList list) {
        return () -> processDLibs(getEventProject(event), (Module) list.getSelectedValue());
    }

    private static boolean dubPathAlreadWarned = false;

    public static void processDLibs(final Project project, @NotNull final Module module) {
        processDLibs(project, module, false, false);
    }

    public static void processDLibs(final Project project,
        @NotNull final Module module, final boolean mostlySilentMode, final boolean buildBefore) {
        ApplicationManager.getApplication().invokeAndWait(new Runnable() {
            @Override
            public void run() {
                final Backgroundable task = new Backgroundable(project,
                    "Updating Dub Libraries") {
                    @Override
                    public void run(@NotNull final ProgressIndicator indicator) {
                        processDLibsImpl(project, module, mostlySilentMode, buildBefore);
                    }
                };
                ProgressManager.getInstance().runProcessWithProgressAsynchronously(
                    task, new BackgroundableProcessIndicator(task));
            }
        }, ModalityState.defaultModalityState());
    }

    public static void processDLibsImpl(final Project project, @NotNull final Module module,
        final boolean mostlySilentMode, final boolean buildBefore) {
        final String prefix = "Unable to process D libraries - ";
        if (project == null) {
            displayError(project, prefix + "No active project.");
            return;
        }
        //todo needs build/fetch before adding libs, also needs to keep track of libs

        // remove all existing libs
//        removeDLibs(module, project);//this is not necessary since intellij filters out duplicate libraries.

        // ask dub for required libs
        final String dubPath = ToolKey.DUB_KEY.getPath();

        //final String groupId = e.getPresentation().getText();
        if (dubPath == null) {
            if (!dubPathAlreadWarned) {
                Notifications.Bus.notify(
                    new Notification(NOTIFICATION_GROUPID, "Process D Libraries",
                        "DUB executable path is empty<br/><a href='configureDLanguageTools'>Configure</a>",
                        NotificationType.WARNING, new DToolsNotificationListener(project)),
                    project);
                dubPathAlreadWarned = true;
            }
            return;
        }

        final DubConfigurationParser dubParser = new DubConfigurationParser(project, dubPath,
            mostlySilentMode);
        if (dubParser.canUseDub()) {
            final List<DubPackage> dependencies = dubParser.getDubPackageDependencies();
            for (final DubPackage pkg : dependencies) {
                final String fullName = pkg.getName() + "-" + pkg.getVersion();
                createLibraryDependency(module, project, fullName, pkg);
            }
        } else {
            LOG.info("not possible to run 'dub describe'");
        }

        if (!mostlySilentMode) {
            Notifications.Bus.notify(
                new Notification(NOTIFICATION_GROUPID, "Process D Libraries",
                    "Added your dub dependency libraries",
                    NotificationType.INFORMATION, new DToolsNotificationListener(project)),
                project);
        }
    }

    private static void createLibraryDependency(final Module module, final Project project,
        final String libraryName, final DubPackage dubPackage) {
        final VirtualFile sources = getSourcesVirtualFile(dubPackage);
        final LibraryTable projectLibraryTable = LibraryTablesRegistrar.getInstance()
            .getLibraryTable(project);
        final LibraryTable.ModifiableModel projectLibraryModel = projectLibraryTable
            .getModifiableModel();

        final Library library = projectLibraryModel.createLibrary(libraryName);
        final Library.ModifiableModel libraryModel = library.getModifiableModel();

        if (sources != null) {
            libraryModel.addRoot(sources, OrderRootType.CLASSES);
            //todo add binary libs/dub.json as well
            ApplicationManager.getApplication()
                .invokeAndWait(() -> ApplicationManager.getApplication().runWriteAction(() -> {
                    libraryModel.commit();
                    projectLibraryModel.commit();
                    for (final Module projectModule : ModuleManager.getInstance(project)
                        .getModules()) {
                        ModuleRootModificationUtil.addDependency(projectModule, library);
                    }

                }));

        }
    }

    private static VirtualFile getSourcesVirtualFile(final DubPackage dubPackage) {
        final String sourcesPathUrl;
        if (SystemInfo.isWindows) {
            sourcesPathUrl = VirtualFileManager.constructUrl(LocalFileSystem.PROTOCOL,
                dubPackage.getPath() + dubPackage.getSourcesDir().replace("/", "\\"));
        } else {
            sourcesPathUrl = VirtualFileManager.constructUrl(LocalFileSystem.PROTOCOL,
                dubPackage.getPath() + dubPackage.getSourcesDir());
        }
        final VirtualFile sources = VirtualFileManager.getInstance()
            .refreshAndFindFileByUrl(sourcesPathUrl);
        if (sources == null) {
            LOG.info("sources not found, fetching them");
            final GeneralCommandLine commandLine = new GeneralCommandLine();
            final String dubBinaryPath = ToolKey.DUB_KEY.getPath();
            if (dubBinaryPath == null) {
                LOG.error("dub executable should be configured");
                return null;
            }
            commandLine.setExePath(dubBinaryPath);
            final ParametersList parametersList = commandLine.getParametersList();
            parametersList.addParametersString("fetch");
            parametersList.addParametersString(dubPackage.getName());
            parametersList.addParametersString("--version=" + dubPackage.getVersion());

            final String dubCommand = commandLine.getCommandLineString();
            final OSProcessHandler process;
            try {
                process = new OSProcessHandler(commandLine.createProcess(), dubCommand);
            } catch (final ExecutionException e) {
                LOG.error(e);
                return null;
            }

            final StringBuilder builder = new StringBuilder();
            final List<String> errors = new ArrayList<>();

            process.addProcessListener(new ProcessAdapter() {
                @Override
                public void onTextAvailable(final ProcessEvent event, final Key outputType) {
                    switch (outputType.toString()) {
                        case "stdout":
                            builder.append(event.getText());
                            break;
                        case "stderr":
                            errors.add(event.getText());
                            break;
                    }
                }
            });

            process.startNotify();
            process.waitFor();

            LOG.debug("output from fetching package:" + builder.toString());

            for (final String error : errors) {
                LOG.error("error while fetching:" + error);
            }

            final Integer exitCode = process.getExitCode();
            if (exitCode != null && exitCode == 0) {
                //try loading again
                return VirtualFileManager.getInstance().refreshAndFindFileByUrl(sourcesPathUrl);
            } else {
                LOG.error("exitcode was no zero for fetching package");
            }
        }
        return sources;
    }

    private static void removeDLibs(final Module module, final Project project) {
        final LibraryTable projectLibraryTable = LibraryTablesRegistrar.getInstance()
            .getLibraryTable(project);
        for (final Library lib : projectLibraryTable.getLibraries()) {
            removeLibraryIfNeeded(module, lib.getName());
        }
    }

    private static void removeLibraryIfNeeded(final Module module, final String libraryName) {
        ApplicationManager.getApplication().assertIsDispatchThread();

        final ModifiableModelsProvider modelsProvider = ModifiableModelsProvider.SERVICE
            .getInstance();
        final ModifiableRootModel model = modelsProvider.getModuleModifiableModel(module);
        final LibraryOrderEntry dLibraryEntry = OrderEntryUtil
            .findLibraryOrderEntry(model, libraryName);
        if (dLibraryEntry != null) {
            ApplicationManager.getApplication().runWriteAction(() -> {
                final Library library = dLibraryEntry.getLibrary();
                if (library != null) {
                    final LibraryTable table = library.getTable();
                    if (table != null) {
                        table.removeLibrary(library);
                        model.removeOrderEntry(dLibraryEntry);
                        modelsProvider.commitModuleModifiableModel(model);
                    }
                } else {
                    modelsProvider.disposeModuleModifiableModel(model);
                }
            });
        } else {
            ApplicationManager.getApplication()
                .runWriteAction(() -> modelsProvider.disposeModuleModifiableModel(model));
        }
    }

    private static void displayError(@NotNull final AnActionEvent e,
        @NotNull final String message) {
        displayError(getEventProject(e), message);
    }

    private static void displayError(final @Nullable Project project,
        @NotNull final String message) {
        //final String groupId = e.getPresentation().getText();
        Notifications.Bus.notify(
            new Notification(NOTIFICATION_GROUPID, "Process D libs", message,
                NotificationType.ERROR),
            project
        );
        LOG.warn(message);
    }

    @Override
    public void update(@NotNull final AnActionEvent e) {
        e.getPresentation().setEnabled(enabled(e));
    }

    @Override
    public void actionPerformed(@NotNull final AnActionEvent e) {
        final String prefix = "Unable to process D libraries - ";
        final Project project = e.getProject();
        if (project == null) {
            displayError(e, prefix + "No active project.");
            return;
        }
        final Collection<Module> modules = DlangModuleType.findModules(project);
        final int size = modules.size();
        if (size == 0) {
            displayError(e, prefix + "No DLanguage modules are used in this project.");
        } else if (size == 1) {
            processDLibs(project, modules.iterator().next());
        } else {
            showModuleChoicePopup(e, project, modules);
        }
    }

}

File: src/main/kotlin/io/github/intellij/dlanguage/actions/ProcessDLibs.kt
package io.github.intellij.dlanguage.actions

import com.intellij.execution.ExecutionException
import com.intellij.execution.configurations.GeneralCommandLine
import com.intellij.execution.process.OSProcessHandler
import com.intellij.execution.process.ProcessAdapter
import com.intellij.execution.process.ProcessEvent
import com.intellij.notification.Notification
import com.intellij.notification.NotificationType
import com.intellij.notification.Notifications
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.ModalityState
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.module.Module
import com.intellij.openapi.module.ModuleManager
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task
import com.intellij.openapi.progress.impl.BackgroundableProcessIndicator
import com.intellij.openapi.project.DumbAware
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.guessProjectDir
import com.intellij.openapi.roots.LibraryOrderEntry
import com.intellij.openapi.roots.ModuleRootModificationUtil
import com.intellij.openapi.roots.OrderRootType
import com.intellij.openapi.roots.libraries.LibraryTablesRegistrar
import com.intellij.openapi.ui.popup.JBPopupFactory
import com.intellij.openapi.util.Key
import com.intellij.openapi.util.SystemInfo
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.VirtualFileManager
import com.intellij.ui.components.JBList
import io.github.intellij.dlanguage.DLanguage
import io.github.intellij.dlanguage.library.DlangLibraryType
import io.github.intellij.dlanguage.module.DlangModuleType
import io.github.intellij.dlanguage.project.DubConfigurationParser
import io.github.intellij.dlanguage.project.DubPackage
import io.github.intellij.dlanguage.settings.ToolKey
import io.github.intellij.dlanguage.utils.DToolsNotificationListener

/**
 * ported from Java on 06/02/18
 *
 * todo: this class needs a rethink. It's currently tied to the user having dub configured. Also, the Action shouldn't
 * need to be called manually. For dub projects we should scan dub.json/sdl for changes or when opening/importing
 * a project (The same way that the Maven/Gradle integration works) and when not using dub we should just
 * detect when the user adds/removes a library.
 */
class ProcessDLibs : AnAction("Process D Libraries", "Processes the D Libraries", DLanguage.Icons.SDK), DumbAware {

    override fun actionPerformed(e: AnActionEvent) {
        val project = e.project
        if (project == null) {
            displayError(e, "Unable to process D libraries - No active project.")
            return
        }
        val modules = DlangModuleType.findModules(project)

        when (modules.size) {
            //0 -> displayError(e, "Unable to process D libraries - No DLanguage modules are used in this project.")
            1 -> processDLibs(project, modules.iterator().next())
            else -> showModuleChoicePopup(e, project, modules)
        }
    }

    /*
    * This method is called repeatedly in UI thread and needs to complete fast. Don't perform any long-running code!
    */
    override fun update(e: AnActionEvent) {
        e.presentation.isEnabled = canRunDub(e)
    }

    companion object {
        private val LOG = Logger.getInstance(ProcessDLibs::class.java)
        private val MENU_PATH = "Tools > Process D Libraries"
        private val NOTIFICATION_GROUPID = "Process D Libs"
        private var dubPathAlreadWarned = false

        /**
         * This method is used from various places, when re-working the way libs are processed this should be marked
         * as deprecated and deleted in a future release
         */
        @JvmStatic
        fun processDLibs(project: Project, module: Module, mostlySilentMode: Boolean, buildBefore: Boolean) {
            ApplicationManager.getApplication().runWriteAction {
                // remove all existing libs from module (removing from module doesn't remove from project):
                removeAllLibrariesFromModule(module)

                removeAllLibrariesFromProject(project)

                // todo: perhaps we can use the custom library tables. Need to find out how they work
                // val libs = LibraryTablesRegistrar.getInstance().customLibraryTables
                // LibraryTablesRegistrar.getInstance().customLibraryTables.clear()

                // used to use the removeDLibs method
                // removeDLibs(module, project); // this is not necessary since intellij filters out duplicate libraries.
            }

            ApplicationManager.getApplication().invokeAndWait({
                val task = object : Task.Backgroundable(project,
                    "Updating Dub Libraries") {
                    override fun run(indicator: ProgressIndicator) {
                        processDLibsImpl(project, module, mostlySilentMode, buildBefore)
                    }
                }
                ProgressManager.getInstance()
                    .runProcessWithProgressAsynchronously(task, BackgroundableProcessIndicator(task))
            }, ModalityState.defaultModalityState())
        }

        private fun processDLibsImpl(project: Project, module: Module,
                                     mostlySilentMode: Boolean, buildBefore: Boolean) {
            //todo needs build/fetch before adding libs, also needs to keep track of libs

            // ask dub for required libs
            val dubPath = ToolKey.DUB_KEY.path

            //final String groupId = e.getPresentation().getText();
            if (dubPath == null) {
                if (!this.dubPathAlreadWarned) {
                    val notification = Notification(
                        NOTIFICATION_GROUPID, "Process D Libraries",
                        "DUB executable path is empty<br/><a href='configureDLanguageTools'>Configure</a>",
                        NotificationType.WARNING
                    )
                    notification.setListener(DToolsNotificationListener(project))

                    Notifications.Bus.notify(notification, project)
                    this.dubPathAlreadWarned = true
                }
                return
            }

            val dubParser = DubConfigurationParser(project, dubPath,
                mostlySilentMode)
            if (dubParser.canUseDub()) {
                val dependencies = dubParser.dubPackageDependencies
                for (pkg in dependencies) {
                    val fullName = pkg.name + "-" + pkg.version
                    createLibraryDependency(module, project, fullName, pkg)
                }
            } else {
                LOG.info("not possible to run 'dub describe'")
            }

            if (!mostlySilentMode) {
                val notification = Notification(
                    NOTIFICATION_GROUPID, "Process D Libraries",
                    "Added your dub dependency libraries",
                    NotificationType.INFORMATION
                )
                notification.setListener(DToolsNotificationListener(project))

                Notifications.Bus.notify(notification, project)
            }
        }

        /*
        * perform some basic checks to see if dub can be run. There should be an active project that contains either
        * a dub.json or dub.sdl file and the path should be set for the dub binary.
        */
        private fun canRunDub(e: AnActionEvent): Boolean {
            val project = AnAction.getEventProject(e) ?: return false
            ToolKey.DUB_KEY.path?.isNotBlank() ?: return false
            val virtualRoot = project.guessProjectDir() ?: return false
            return virtualRoot.findChild("dub.json")?.exists() ?:
                            virtualRoot.findChild("dub.sdl")?.exists() ?:
                            false
        }

        private fun showModuleChoicePopup(e: AnActionEvent, project: Project, modules: Collection<Module>) {
            val list = JBList(JBList.createDefaultListModel<Module>(*modules.toTypedArray()))
            val popup = JBPopupFactory.getInstance()
                .createListPopupBuilder(list)
                .setTitle("Process D libraries for module")
                .setItemChoosenCallback { processDLibs(AnAction.getEventProject(e), list.selectedValue as Module) }
                .createPopup()
            popup.showCenteredInCurrentWindow(project)
        }

        private fun processDLibs(project: Project?, module: Module) {
            if (project == null) {
                LOG.warn("Unable to process D libraries - No active D project.")
                //displayError(project, "Unable to process D libraries - No active D project.");
                return
            }
            processDLibs(project, module, false, false)
        }

        private fun createLibraryDependency(module: Module, project: Project,
                                            libraryName: String, dubPackage: DubPackage) {
            getSourcesVirtualFile(dubPackage)?.let { sources ->
                val projectLibraryTable = LibraryTablesRegistrar.getInstance().getLibraryTable(project)
                val projectLibraryModel = projectLibraryTable.modifiableModel

                //val uniqueLibName = LibraryEditingUtil.suggestNewLibraryName(projectLibraryModel, libraryName) // not needed as we remove clashes
                val library = projectLibraryModel.createLibrary(libraryName, DlangLibraryType.DLANG_LIBRARY) // todo: use the ProjectModelExternalSource arg

                val libraryModel = library.modifiableModel

                libraryModel.addRoot(sources, OrderRootType.CLASSES) // todo: Use our own OrderRootType, such as LibFileRootType.getInstance()

                //todo add binary libs/dub.json as well
                ApplicationManager.getApplication()
                    .invokeAndWait {
                        ApplicationManager.getApplication().runWriteAction {
                            libraryModel.commit()
                            projectLibraryModel.commit()
                            for (projectModule in ModuleManager.getInstance(project).modules) {
                                // todo: should we be doing ModuleRootModificationUtil.addModuleLibrary()
                                ModuleRootModificationUtil.addDependency(projectModule, library)
                                LOG.debug("added dependency ${library.presentableName} to module '${projectModule.name}'")
                            }

                        }
                    }

            }
        }

        // todo: This method needs breaking down. Also, a dub package can have multiple source directories
        private fun getSourcesVirtualFile(dubPackage: DubPackage): VirtualFile? {
            // for now I'm putting this hideous kludge in which makes it work they way it always has
            val srcDir = if(dubPackage.sourcesDirs.isNotEmpty()) {
                if(dubPackage.sourcesDirs.size > 1) {
                    LOG.warn("it's likely that we're not processing ${dubPackage.name} correctly. ${dubPackage.sourcesDirs.size} source folders found")
                }
                dubPackage.sourcesDirs[0]
            } else "source"

            val sourcesPathUrl = if (SystemInfo.isWindows) {
                VirtualFileManager.constructUrl(LocalFileSystem.PROTOCOL, dubPackage.path + srcDir.replace("/", "\\"))
            } else {
                VirtualFileManager.constructUrl(LocalFileSystem.PROTOCOL, dubPackage.path + srcDir)
            }

            return VirtualFileManager.getInstance()
                .refreshAndFindFileByUrl(sourcesPathUrl) ?: return this.dubFetch(dubPackage, sourcesPathUrl)
        }

        private fun dubFetch(dubPackage: DubPackage, sourcesPathUrl: String): VirtualFile? {
            LOG.info("sources not found, fetching them")
            val commandLine = GeneralCommandLine()
            val dubBinaryPath = ToolKey.DUB_KEY.path
            if (dubBinaryPath == null) {
                LOG.error("dub executable should be configured")
                return null
            }
            commandLine.exePath = dubBinaryPath
            val parametersList = commandLine.parametersList
            parametersList.addParametersString("fetch")
            // todo: add some logic here to choose between "dub fetch <package> --version=x.y.z" and "dub fetch <package>@x.y.z"
            // see: https://github.com/dlang/dub/issues/2028
            // if dub is version 1.23 or over:
            parametersList.addParametersString("${dubPackage.name}@${dubPackage.version}")
            // else use old syntax:
            // parametersList.addParametersString(dubPackage.name)
            // parametersList.addParametersString("--version=${dubPackage.version}")

            val dubCommand = commandLine.commandLineString
            val process: OSProcessHandler
            try {
                process = OSProcessHandler(commandLine.createProcess(), dubCommand)
            } catch (e: ExecutionException) {
                LOG.error(e)
                return null
            }

            val builder = StringBuilder()
            val errors = ArrayList<String>()

            process.addProcessListener(object : ProcessAdapter() {
                override fun onTextAvailable(event: ProcessEvent, outputType: Key<*>) {
                    when (outputType.toString()) {
                        "stdout" -> builder.append(event.text)
                        "stderr" -> errors.add(event.text)
                    }
                }
            })

            process.startNotify()
            process.waitFor()

            LOG.debug("output from fetching package: $builder")

            errors.forEach {
                LOG.warn("Error when running '${dubCommand}' : $it")
            }

            val exitCode = process.exitCode
            if (exitCode != null && exitCode == 0) {
                //try loading again
                return VirtualFileManager.getInstance().refreshAndFindFileByUrl(sourcesPathUrl)
            } else {
                LOG.error("exitcode was no zero for fetching package '${dubPackage.name}' version '${dubPackage.version}'")
            }
            return null
        }

        private fun removeAllLibrariesFromModule(module: Module) {
            ApplicationManager.getApplication().assertWriteAccessAllowed()

            LOG.debug("Clearing libraries from module '${module.name}'")

            ModuleRootModificationUtil.updateModel(module) { model ->
                model.orderEntries
                    .filterIsInstance<LibraryOrderEntry>()
                    .forEach { lib ->
                        model.removeOrderEntry(lib)
                        LOG.debug("removed ${lib.presentableName} from from module '${module.name}'")
                    }
            }
        }

        private fun removeAllLibrariesFromProject(project: Project) {
            ApplicationManager.getApplication().assertWriteAccessAllowed()

            LOG.debug("Clearing libraries from project '${project.name}'")

            LibraryTablesRegistrar.getInstance()
                .getLibraryTableByLevel(LibraryTablesRegistrar.PROJECT_LEVEL, project)?.let { libraryTable ->
                    libraryTable.modifiableModel.libraries.iterator().forEach { lib ->
                        libraryTable.removeLibrary(lib)
                        LOG.debug("removed ${lib.presentableName} from from project '${project.name}'")
                    }
                }
        }

//    private fun removeDLibs(module: Module, project: Project) {
//        val projectLibraryTable = LibraryTablesRegistrar.getInstance()
//            .getLibraryTable(project)
//        for (lib in projectLibraryTable.libraries) {
//            removeLibraryIfNeeded(module, lib.name)
//        }
//    }

//        private fun removeLibraryIfNeeded(module: Module, libraryName: String?) {
//            ApplicationManager.getApplication().assertIsDispatchThread()
//
//            val modelsProvider = ModifiableModelsProvider.SERVICE.getInstance()
//            val model = modelsProvider.getModuleModifiableModel(module)
//            val dLibraryEntry = OrderEntryUtil.findLibraryOrderEntry(model, libraryName!!)
//
//            if (dLibraryEntry != null) {
//                ApplicationManager.getApplication().runWriteAction {
//                    val library = dLibraryEntry.library
//                    if (library != null) {
//                        val table = library.table
//                        if (table != null) {
//                            table.removeLibrary(library)
//                            model.removeOrderEntry(dLibraryEntry)
//                            modelsProvider.commitModuleModifiableModel(model)
//                        }
//                    } else {
//                        modelsProvider.disposeModuleModifiableModel(model)
//                    }
//                }
//            } else {
//                ApplicationManager.getApplication()
//                    .runWriteAction { modelsProvider.disposeModuleModifiableModel(model) }
//            }
//        }

        private fun displayError(e: AnActionEvent, message: String) {
            displayError(AnAction.getEventProject(e), message)
        }

        private fun displayError(project: Project?, message: String) {
            //final String groupId = e.getPresentation().getText();
            Notifications.Bus.notify(
                Notification(NOTIFICATION_GROUPID, "Process D libs", message, NotificationType.ERROR),
                project
            )
            LOG.warn(message)
        }
    }

}
